pkg/controller/servicecidrs/servicecidrs_controller.go

```diff
@@ -0,0 +1,500 @@
+	workerLoopPeriod time.Duration
+
+	// tree store the ServiceCIDRs names associated to each
+	muTree sync.Mutex
+	treeV4 *iptree.Tree[cidrSet]
+	treeV6 *iptree.Tree[cidrSet]
+}
+
+// Run will not return until stopCh is closed.
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.eventBroadcaster.StartStructuredLogging(0)
+	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.client.CoreV1().Events("")})
+	defer c.eventBroadcaster.Shutdown()
+
+	logger := klog.FromContext(ctx)
+
+	logger.Info("Starting", "controller", controllerName)
+	defer logger.Info("Shutting down", "controller", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, ctx.Done(), c.serviceCIDRsSynced, c.ipAddressSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.UntilWithContext(ctx, c.worker, c.workerLoopPeriod)
+	}
+	<-ctx.Done()
+}
+
+func (c *Controller) addServiceCIDR(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+func (c *Controller) updateServiceCIDR(oldObj, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+// deleteServiceCIDR
+func (c *Controller) deleteServiceCIDR(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+// addIPAddress may block a ServiceCIDR deletion
+func (c *Controller) addIPAddress(obj interface{}) {
+	ip, ok := obj.(*networkingapiv1alpha1.IPAddress)
+	if !ok {
+		return
+	}
+
+	for _, cidr := range c.ipToCidrs(ip) {
+		c.queue.Add(cidr)
+	}
+}
+
+// deleteIPAddress may unblock a ServiceCIDR deletion
+func (c *Controller) deleteIPAddress(obj interface{}) {
+	ip, ok := obj.(*networkingapiv1alpha1.IPAddress)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return
+		}
+		ip, ok = tombstone.Obj.(*networkingapiv1alpha1.IPAddress)
+		if !ok {
+			return
+		}
+	}
+
+	for _, cidr := range c.ipToCidrs(ip) {
+		c.queue.Add(cidr)
+	}
+}
+
+// ipToCidrs, given an IPAddress return the ServiceCIDRs with the longest prefix match
+// that contains the IP, as it may block or be blocking the deletion of the ServiceCIDR
+func (c *Controller) ipToCidrs(ip *networkingapiv1alpha1.IPAddress) []string {
+	// only process IPs managed by the kube-apiserver
+	managedBy, ok := ip.Labels[networkingapiv1alpha1.LabelManagedBy]
+	if !ok || managedBy != ipallocator.ControllerName {
+		return []string{}
+	}
+
+	ipFamily, ok := ip.Labels[networkingapiv1alpha1.LabelIPAddressFamily]
+	if !ok {
+		return []string{}
+	}
+
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+
+	// the only serviceCIDRs that can be affected by an IPAddress deletiong
+	// are the ones with the longest prefix match.
+	var cidrs cidrSet
+	if ipFamily == string(v1.IPv4Protocol) {
+		_, cidrs, ok = c.treeV4.LongestPrefixMatch(ip.Name + "/32")
```

Comment:
My earlier comment about .0 and broadcast addresses applies here and is more explicitly a bug; if you have allocators for `192.168.1.0/24` and `192.168.0.0/16`, then this will say that `192.168.1.0` belongs to `192.168.1.0/24`, but it actually belongs to `192.168.0.0/16`.

Reply:
I had a discussion with @thockin some time ago I can't find, but there was also an RFC explaining that  192.168.1.0 or 192.168.1.255 are a perfect valid address in 192.168.1.0/16 , meanwhile you don't use a broadcast medium as ethernet of course

Reply:
Getting rid of the two unallocated IPs per CIDR seems like a great idea, if we think it's safe.

Reply:
nah, you are right , the IPAllocator does implement the same logic we always had, we don't allocate the .0 and the broadcast only in IPv6, is a very subtle bug, thanks for catching

Reply:
fixed